### ISSUE TITLE
Empty string returned during select when column is NULL

### ISSUE THREAD

**Issue by mateuscb** [sidorares/node-mysql2#637]
I have the following scenario:

- Both big number configs: `{supportBigNumbers: true, bigNumberStrings: true}`
- Table with a `bigint(20) unsigned` column that is nullable.

When I performed a `select` I expected to get back `null` but instead I get back `''` for that column.

I [modified the one of the tests](https://github.com/sidorares/node-mysql2/compare/master...mateuscb:patch-1) to ensure that the issue was reproachable. To my surprise, the return was different when using  `conn.query()` vs `conn.execute()`.  `execute` returned `null` as I expected, but `query` returned `''`.

Is this expected or should they both return the same thing? If they are to return the same thing, should it be `null` or `''` (I personally thing it should be `null`).

**Comment by sidorares** [comment]
Thanks, looks like it's a bug.
Query and execute results are serialised differently: query uses 'text protocol' where '12345' number is sent as string: length-coded prefix, then data as string. For `execute` numbers are encoded as numbers ( 16 / 32 / 64 bit )

[readLengthCodedNumber](https://github.com/sidorares/node-mysql2/blob/c9e592643a2688de18870c1b75ff5fde1467c346/lib/packets/packet.js#L152-L209) reads variable length encoded number, and one of results can be 'null'.

[This function](https://github.com/sidorares/node-mysql2/blob/c9e592643a2688de18870c1b75ff5fde1467c346/lib/packets/packet.js#L667) incorrectly returns space when value is null, we need to add check similar to [this code](https://github.com/sidorares/node-mysql2/blob/c9e592643a2688de18870c1b75ff5fde1467c346/lib/packets/packet.js#L345-L359)

**Comment by sidorares** [comment]
Probably fix is to make `parseLengthCodedInt()` to look like this:

```js
Packet.prototype.parseLengthCodedIntString = function() {
  return this..readLengthCodedString('binary');
};
```

**Comment by sidorares** [comment]
I'd also review https://github.com/sidorares/node-mysql2/blob/c9e592643a2688de18870c1b75ff5fde1467c346/lib/packets/packet.js#L663-L677 for how they handle `null` value in result

Can you review first @mateuscb ?

**Comment by mateuscb** [comment]
Thanks, that's a lot of good stuff. I'll see if I can figure out how to solve this. 

What did you mean by "can you review first"?

**Comment by mateuscb** [comment]
That makes sense. I won't be able to look at it today. I was planning to look at it tomorrow. If you don't get to it today, I'll def check it out tomorrow morning (UTC-7).

**Comment by mateuscb** [comment]
Took me a while to kind of wrap my mind around the entire code base. I've come across what I think may be another issue (or perhaps I'm doing something wrong). But if I comment out all but the last ` testExecute(true, true, bigNums_execute_bnStringsTrueTrue);` and put a `console.log(bigNumberStrings)` in `Packet.prototype.readLengthCodedNumber = function(bigNumberStrings, signed)` it's always false. I was debugging it to try to figure out why and think it has something to do with how `command/execute.js` deals with the `this.options`.  It seems to grab the `_executeOptions` correctly but the connection it uses (`this._connection`) never gets updated with `{supportBigNumbers:true,bigNumberStrings:true}`. Does that sound right? Or did am I doing something wrong.

If I update the `common.createConnection();` to include `{supportBigNumbers:true,bigNumberStrings:true}` then `readLengthCodedNumber` has `bigNumberStrings==true`.

**Comment by mateuscb** [comment]
Per your suggestion, I updated `parseLengthCodedIntString` to:

```
Packet.prototype.parseLengthCodedIntString = function() {
  return this.readLengthCodedString('binary');
};
```

That seems to fix the issue.  I also looked into the other three methods you mentioned (`parseLengthCodedInt`, `parseLengthCodedInt`, `parseLengthCodedFloat`) and the methods they call (`parseIntNoBigCheck`, `parseInt`, `parseFloat`) all check for `len === null`. So, it looks like we should be good for those methods.

Here is the commit that added the `len == 'undefined'` in `readString()` (maybe it should also have `len === null`?):
https://github.com/sidorares/node-mysql2/commit/99a8299ed1eb60eee0692175b61d26ba606d9056#diff-e5bb9d341e475547ebd4eb4462b097d7

Here is the commit where `parseLengthCodedIntString()` was added:
https://github.com/sidorares/node-mysql2/commit/6e3e91dadcc6aaa0e237c55a2eb945653a7f0515#diff-e5bb9d341e475547ebd4eb4462b097d7

I've updated my branch, but not sure if that small change will have side effects I'm not aware of.

Thoughts?
